libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,139 @@
+#include <stdbool.h>
+
+#include "adc.h"
+#include "log.h"
+#include "stm32f0xx.h"
+
+typedef struct ADCInterrupt {
+  ADCCallback callback;
+  void *context;
+  uint16_t reading;
+} ADCInterrupt;
+
+typedef struct ADCStatus {
+  ADCChannel current_channel;
+  bool continuous;
+} ADCStatus;
+
+static ADCInterrupt s_adc_interrupts[NUM_ADC_CHANNEL];
+static ADCStatus s_adc_status;
+
+static void prv_adc_conversion_sequence() {
+  s_adc_status.current_channel = 0;
+  ADC_StartOfConversion(ADC1);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_EOSEQ)) {}
+}
+
+static uint32_t prv_get_channel_select() {
+  return ADC1->CHSELR;
+}
+
+void adc_init(ADCMode adc_mode) {
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+  ADC_InitTypeDef adc_settings = {
+    ADC_Resolution_12b,
```

Comment:
We generally use designated initializers to be more explicit about what values we're setting.